### ISSUE TITLE
Is the way how you handle $CROSS_COMPILE and $CC really correct?

### ISSUE THREAD

**Issue by CodingMarkus** [openssl/openssl#24460]
When trying to cross compile lots of code, you typically have an environment where `$CROSS_COMPILE` and `$CC` are both set. E.g. I tried to cross-compile OpenSSL with [emscripten](https://docs.huihoo.com/emscripten/emscripten-site/docs/porting/connecting_cpp_and_javascript/embind.html) and this caused the following issue:

emscripten will set both variables as follows

```
CROSS_COMPILE=/full/path/to/emscripten/em
CC=/full/path/to/emscripten/emcc
```

Most autoconf packages either don't know anything about cross compiling and just use `$CC`, which will work fine, or they are aware of `$CROSS_COMPILE` and will set

```
CC=${CROSS_COMPILE}cc
```

which will work as well.

But OpenSSL sets

```
CC=${CROSS_COMPILE}${CC}
```

which will result in 

```
CC=/full/path/to/emscripten/em/full/path/to/emscripten/emcc
```

And that will of course fail and I wonder if that behavior is correct. Are you really supposed to append `$CC` to `$CROSS_COMPILE` and not just append `cc` (or `cpp` or `ld`) at the end?

I always thought that you either use `$CC` when set and only if not set, but `$CROSS_COMPILE` is set, you would use `$CROSS_COMPILE` to begin with and never use both, as setting `$CC` should always override which compile is being used. At least if `$CC` has an absolute path I'd always ignore `$CROSS_COMPILE`.

**Comment by t8m** [comment]
Yeah, if we get CC from environment with an absolute path, we definitely should not prepend CROSS_COMPILE.

**Comment by CodingMarkus** [comment]
Just FYI, I also created an issue with emscripten, as I wasn't sure if their way of setting up `$CROSS_COMPILE` and `$CC` is the expected way of doing things. 

https://github.com/emscripten-core/emscripten/issues/21969

I doubt there are any "official rules" for cross compiling, but maybe there are unofficial de-facto standards that have established over the years. 

So just in case you think their way of setting variables might not be the right way, they would also be willing to change it and accept a patch I could give them. I'm just trying to figure out what the right thing to do is. It's always a bit tricky when things just don't work, but you can't be sure who's to blame, and sometimes the answer is "both" and sometimes it's actually "nobody".

I think the proposed logic regarding absolute paths makes perfect sense, but if you have any additional preferences on how to set up a better cross-compiler environment, I'd be happy to take your input on that other issue.

**Comment by rollebolle** [comment]
The way OpenSSL does it is clearly wrong and goes against de-facto standard, and is a headache to use when cross-compiling for example from Yocto. In cross-compilation $CC contains the compiler WITH the target triplet, and additional prefixing with CROSS_COMPILE will result in errors

**Comment by levitte** [comment]
The idea with `$CROSS_COMPILE` is (and has been all the way since 2009) to be a prefix, so that you wouldn't have to make any other changes (i.e. no need to setup `$CC`).

Judging from a quick readup of https://docs.huihoo.com/emscripten/emscripten-site/docs/getting_started/downloads.html, it seems emscripten provides a full toolchain, which tells me the example setup for OpenSSL should be something like this:

    CROSS_COMPILE=/full/path/to/emscripten/em
    CC=cc

Do note that other tools in the toolchain will be prefixed with `$(CROSS_COMPILE)` as well, for example `ar`, so this assumes that there is also a `/full/path/to/emscripten/emar`.  If this is not the case, *then it's better to make `$CROSS_COMPILE` empty*, and set these variables to the full path of each tool:

- `CC` (should 
- `CXX`
- `AR`
- `RANLIB`
- `RC`

(you find them with `grep CROSS_COMPILE` in the Makefile that OpenSSL's Configure builds)

**Comment by rollebolle** [comment]
This is  an example of how OpenEmbedded/Yocto defines it in the file you source before using their SDK:

```
export CC="x86_64-oe-linux-gcc  -m64 -march=core2 -mtune=core2 -msse3 -mfpmath=sse --sysroot=$SDKTARGETSYSROOT"
export CROSS_COMPILE=x86_64-oe-linux-
```
I have compiled many other packages using this file, and OpenSSL is the only one that uses the $CROSS_COMPILE-$CC command line (leading to an error and that we have to temporarily clear the CROSS_COMPILE variable). 

Of course there is no set standard for this, but OpenEmbedded is a very large and popular environment for cross compilation, so it might be a good idea to be aliged with them (and emscripten if I understand correctly)

**Comment by topiolli** [comment]
Any progress on this? It causes me constant headache because OpenSSL pops up as a transitive dependency all the time, and I cannot patch all the installers. Every other dependency I have needed uses either `$CC` or `${CROSS_COMPILE}cc` (and similarly for the other tools). My work-around is to create symlinks so that OpenSSL can use `${CROSS_COMPILE}${CC}`, but that's ugly.

**Comment by t8m** [comment]
Please note we are looking for community help implementing this issue. Please comment if you are interested in implementing this feature or if you have reasons why this feature is important enough so we should consider implementing it with higher priority than other requested features. The issue will be closed after 2 months otherwise.

**Comment by t8m** [comment]
This issue was closed because it was marked with the help wanted label and there was no interest in implementing this issue or any other activity leading to implementing or re-prioritizing it for more than 2 months.
